### PR TITLE
Fix item-list index issues while sorting/filtering

### PR DESCRIPTION
**Pull request by yebrahim**
When sorting/filtering, the selected items' real indices should be distinguished from their render index. This PR fixes a bug where this wasn't the element's behavior, and the element returned the render indices of the items selected.

Additionally, this PR also fixes the behavior of Select/Unselect All to only work on the currently visible (filtered) items.

Fixes https://github.com/googledatalab/datalab/issues/1857.

### CODE REVIEW

File: sources/web/datalab/polymer/components/item-list/item-list.ts [hunk 1]
```diff
@@ -451,63 +469,84 @@ class ItemListElement extends Polymer.Element {
...
+      this._unselectAllDisplayedItems();
     }
   }
 
   /**
    * On row click, checks the click target, if it's the checkbox, adds it to
    * the selected rows, otherwise selects it only.
+   * Note the distinction between displayIndex, which is the index of the item
+   * in the rendered list, and realIndex, which is the index of the item in the
+   * original list that was submitte to the item-list element. These might be
```
Comment (jimmc) [inline sources/web/datalab/polymer/components/item-list/item-list.ts]: typo: submitted

File: sources/web/datalab/polymer/test/item-list-test.ts [hunk 2]
```diff
@@ -683,6 +699,26 @@ describe('<item-list>', () => {
         assert(columns[0].innerText === testFixture.rows[sortedOrder[i]].columns[0].toString());
       }
     });
+
+    it('returns the correct selectedIndices result matching clicked items', () => {
```
Comment (jimmc) [inline sources/web/datalab/polymer/test/item-list-test.ts]: Should there be a test that includes sorting?

Comment (yebrahim) [reply sources/web/datalab/polymer/test/item-list-test.ts]: This one does. It sorts the list first, reversing the order of the items 3 -> 2 -> 1, then selects the first one and makes sure its selected index is 2 not 0.

File: sources/web/datalab/polymer/components/item-list/item-list.ts (hunk 1 above)
Comment (yebrahim) [reply sources/web/datalab/polymer/components/item-list/item-list.ts]: Done.

File: sources/web/datalab/polymer/test/item-list-test.ts (hunk 2 above)
Comment (yebrahim) [reply sources/web/datalab/polymer/test/item-list-test.ts]: I can add one more for filtering.

**Pull request merged**